readthedocs/search/parsers.py

```diff
@@ -123,6 +123,59 @@ def _parse_sections(self, title, body):
         We can have pages that have content before the first title or that don't have a title,
         we index that content first under the title of the original page.
         """
+
+        # All terms in dls are treated as sections.
+        # After indexing them, we remove them from the body such that their contents aren't
+        # repeated in the following indexing
+        dls = body.css("dl")
+        seen_dls = []
+        for dl in dls:
+            if dl in seen_dls:
+                continue
+            for inner_dl in dl.css("dl"):
+                seen_dls.append(inner_dl)
+
+            # Select all dts with id defined
+            dts = dl.css('dt[id]:not([id=""])')
+
+            # https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dt
+            # multiple <dt> elements in a row indicate several terms that are
+            # all defined by the immediate next <dd> element.
+            for dt in dts:
+                try:
+                    title, _id = self._parse_dt(dt)
+                    # https://developer.mozilla.org/en-US/docs/Web/CSS/General_sibling_combinator
+                    dd = dt.css_first("dt ~ dd")
+                    # We only index a dt with an id attribute and an accompanying dd
+                    if not dd or not _id:
+                        continue
+
+                    # Create a copy of the node to avoid manipulating the
+                    # data structure that we're iterating over
+                    dd_copy = HTMLParser(dd.html).body.child
+
+                    # Remove all nested domains from dd_copy.
+                    # They are already parsed separately.
+                    nested_domains = dd_copy.css("dl")
+                    for node in nested_domains:
+                        if dd_copy != node:
+                            node.decompose()
+
+                    # The content of the <dt> section is the content of the accompanying <dd>
+                    content = self._parse_content(dd_copy.text())
+
+                    yield {
+                        "id": _id,
+                        "title": title,
+                        "content": content,
+                    }
+                except Exception:
+                    log.info("Unable to index dt section.", exc_info=True)
+
+            # Remove the <dl> node
+            # There isn't a clear indication if this behavior is DFS or BFS
```

Comment:
Didn't you figure out that this is BFS? Probably worth updating the comment.

But also, if this is BFS, we will be deleting the nested lists here too, I think.

Reply:
I _experience_ that it traverses like BFS... however I did not read the implementation :) So I still think we're safest in not assuming the order here?

Reply:
Yeah, I'm +1 on not assuming the order, so we need to delete all indexed nodes at the end of the loop, not inside.

Reply:
Yes, :100: 

I've changed it locally and it actually _does_ affect indexing on test data :+1:

Reply:
This is now finally done in the most accurate manner by removing exactly the `<dt>, <dd>` elements that are indexed :+1: